flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractEmbeddedPythonFunctionOperator.java

```diff
@@ -0,0 +1,176 @@
+            lock.lockInterruptibly();
+
+            JobID jobId = getRuntimeContext().getJobId();
+            Tuple2<String, Integer> dirAndNums;
+
+            if (workingDirectories.containsKey(jobId)) {
+                dirAndNums = workingDirectories.get(jobId);
+            } else {
+                dirAndNums = Tuple2.of(null, 0);
+                workingDirectories.put(jobId, dirAndNums);
+            }
+            dirAndNums.f1 += 1;
+
+            if (dirAndNums.f0 == null) {
+                // get current directory.
+                interpreter.exec("import os;cwd = os.getcwd();");
+                dirAndNums.f0 = interpreter.get("cwd", String.class);
+                String workingDirectory = env.get(PYTHON_WORKING_DIR);
+                // set working directory
+                interpreter.exec(String.format("import os;os.chdir('%s')", workingDirectory));
+            }
+
+            lock.unlock();
+        }
+
+        if (interpreterConfig
+                .getExecType()
+                .equals(PythonInterpreterConfig.ExecType.SUB_INTERPRETER)) {
+            LOG.info("Create Operation in multi interpreters.");
+            createOperationInMultiInterpreters(pythonConfig.getPythonExec(), env);
+        } else {
+            LOG.info("Create Operation in multi threads.");
+            createOperationInMultiThread();
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        try {
+            JobID jobId = getRuntimeContext().getJobId();
+            if (workingDirectories.containsKey(jobId)) {
+                lock.lockInterruptibly();
+                Tuple2<String, Integer> dirAndNums = workingDirectories.get(jobId);
+                dirAndNums.f1 -= 1;
+                if (dirAndNums.f1 == 0) {
+                    // change to previous working directory.
+                    interpreter.exec(String.format("import os;os.chdir('%s')", dirAndNums.f0));
+                }
+                lock.unlock();
+            }
+
+            if (interpreter != null) {
+                interpreter.close();
+            }
+
+            if (pythonEnvironmentManager != null) {
+                pythonEnvironmentManager.close();
+            }
+        } finally {
+            super.close();
+        }
+    }
+
+    @Override
+    protected EmbeddedPythonEnvironmentManager createPythonEnvironmentManager() {
+        PythonDependencyInfo dependencyInfo =
+                PythonDependencyInfo.create(
+                        pythonConfig, getRuntimeContext().getDistributedCache());
+        return new EmbeddedPythonEnvironmentManager(
+                dependencyInfo,
+                getContainingTask().getEnvironment().getTaskManagerInfo().getTmpDirectories(),
+                new HashMap<>(System.getenv()),
+                getRuntimeContext().getJobId());
+    }
+
+    /** Creates the operation in multi-thread mode. */
+    public abstract void createOperationInMultiThread();
```

Comment:
What about rename it to openPythonInterpreter and merge createOperationInMultiThread and createOperationInMultiInterpreters into one method